File: src/edu/gbcg/dbInteraction/dbcreator/reddit/Facilitator.java
package edu.gbcg.dbInteraction.dbcreator.reddit;

import edu.gbcg.configs.Finals;
import edu.gbcg.dbInteraction.DBCommon;
import edu.gbcg.dbInteraction.dbcreator.IndexWorker;
import edu.gbcg.utils.FileUtils;
import edu.gbcg.utils.TSL;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.sql.Connection;
import java.util.ArrayList;
import java.util.List;

public abstract class Facilitator {
    protected List<String> DBAbsolutePaths;     // Path to the DBs once the exist
    protected List<String> DBDirectoryPaths;    // Path to the directory / directories that hold the DB shards
    protected List<String> columnNames;         // Names of the columns in the DB
    protected List<String> dataTypes;           // Type of data stored in the DB columns
    protected List<String> DBPaths;             // Paths to the DBs when they don't yet exist
    protected List<String> jsonAbsolutePaths;   // Paths to the json files
    protected List<String> jsonKeysOfInterest;  // JSON keys we care about grabbing for the DB
    protected String tableName;                 // The name of the table in the DB

    public Facilitator(){
        this.DBAbsolutePaths        = getDBAbsolutePaths();
        this.DBDirectoryPaths       = getDBDirectoryPaths();
        this.columnNames            = getColumnNames();
        this.dataTypes              = getDataTypes();
        this.DBPaths                = buildDBPaths();
        this.jsonAbsolutePaths      = getJsonAbsolutePaths();
        this.jsonKeysOfInterest     = getJsonKeysOfInterest();
        this.tableName              = getTableName();
    }

    // Used when the DBs don't exist, build the path to the DBs
    protected abstract List<String> buildDBPaths();
    protected abstract List<String> getJsonAbsolutePaths();
    protected abstract List<String> getDBAbsolutePaths();
    protected abstract List<String> getDBDirectoryPaths();
    protected abstract List<String> getDataTypes();
    protected abstract List<JsonPusher> populateJsonWorkers();
    protected abstract List<String> getJsonKeysOfInterest();
    protected abstract String getTableName();
    protected abstract void createIndices();
    protected abstract List<String> getColumnNames();

    public void createDBs(){
        // Check if the all the DBs exist. Note, this is 100% but it's good enough for my uses
        if(this.DBAbsolutePaths == null)
            this.DBAbsolutePaths = new ArrayList<>();
        boolean dbs_exist = this.DBAbsolutePaths.size() == Finals.DB_SHARD_NUM;

        // Early exist if the DBs exist and we don't want to start fresh
        if(dbs_exist && !Finals.START_FRESH)
            return;

        // The DBs exist but we want to start fresh, get rid of them
        if(dbs_exist && Finals.START_FRESH){
            String sql = "drop table if exists " + this.tableName + ";";
            for(String dbPath : this.DBAbsolutePaths)
                DBCommon.delete(dbPath, sql);
        }

        // The DBs don't exist, create them
        if(!dbs_exist){
            // Create the directories that hold the DBs
            for(String path : DBDirectoryPaths)
                FileUtils.get().checkAndCreateDir(path);
            // Build the paths to the DBs so they can be created
            for(String path : DBPaths){
                Connection conn = DBCommon.connect(path);
                DBCommon.disconnect(conn);
            }
            // Now they exist, populate the path data structure
            DBAbsolutePaths = getDBAbsolutePaths();
        }

        // Create the table schema
        StringBuilder sb = new StringBuilder();
        sb.append("create table if not exists "+this.tableName+"(");
        for(int i = 0; i < this.columnNames.size(); ++i){
            sb.append(this.columnNames.get(i));
            sb.append(this.dataTypes.get(i));
        }
        sb.append(");");
        String sql = sb.toString();

        // Create the table in the DB shards
        for(String DB : DBAbsolutePaths)
            DBCommon.insert(DB, sql);
    }

    public void pushJSONDataIntoDBs(){
        // Early exit if we're not pushing data
        if(!Finals.START_FRESH) return;

        if(DBAbsolutePaths == null || DBAbsolutePaths.isEmpty())
            DBAbsolutePaths = getDBAbsolutePaths();
        if(jsonAbsolutePaths == null || jsonAbsolutePaths.isEmpty())
            jsonAbsolutePaths = getJsonAbsolutePaths();

        // For each json file, read it line by line, while reading start processing the data
        // Each iteration of the loop adds a line to a new worker thread to evenly shard the data across all DB shards
        for(String json : jsonAbsolutePaths){
            File f = new File(json);
            TSL.get().info("Reading "+f.getName());

            BufferedReader br = null;
            // How many lines to read before writing to a DB shard
            int dbDumpLimit = Finals.DB_SHARD_NUM * Finals.DB_BATCH_LIMIT;
            int lineReadCounter = 0;
            int whichWorker = 0;

            List<List<String>> linesList = new ArrayList<>();
            for(int j = 0; j < Finals.DB_SHARD_NUM; ++j)
                linesList.add(new ArrayList<>());

            try{
                br = new BufferedReader(new FileReader(json));
                String line = br.readLine();
                while(line != null){
                    ++lineReadCounter;
                    linesList.get(whichWorker).add(line);

                    // Increment the worker number so we're evenly distributing lines to the workers
                    ++whichWorker;
                    if(whichWorker >= Finals.DB_SHARD_NUM)
                        whichWorker = 0;

                    // Limit before dumping data to the DB, start the threads and perform the dump
                    if(lineReadCounter >= dbDumpLimit){
                        letWorkersFly(linesList);

                        // Reset the trackers and clear the lines list to recover the memory from 5000 lines of JSON
                        lineReadCounter = 0;
                        linesList.clear();

                        // Give the linesList new ArrayLists to store the lines
                        for(int j = 0; j < Finals.DB_SHARD_NUM; ++j)
                            linesList.add(new ArrayList<>());
                    }
                    // Read a line
                    line = br.readLine();
                }

                // There could be leftover json lines that don't get push due to not meeting the dbDumpLimit amount
                // of lines, start up the workers again and push the remaining data
                letWorkersFly(linesList);
            }
            catch(IOException e){
                TSL.get().err("Facilitator.pushJSONDataIntoDBs IOException");
            }
            finally{
                if(br != null){
                    try{
                        br.close();
                    }
                    catch(IOException e){
                        TSL.get().err("Facilitator.pushJSONDataIntoDBs br.close IOException");
                    }
                }
            }
        }
        createIndices();
    }

    private void letWorkersFly(List<List<String>> lines){
        List<JsonPusher> workers = populateJsonWorkers();
        // Give the workers the data they need
        for(int i = 0; i < Finals.DB_SHARD_NUM; ++i){
            workers.get(i).setDB(DBAbsolutePaths.get(i));
            workers.get(i).setJSON(lines.get(i));
            workers.get(i).setColumns(columnNames);
            workers.get(i).setTableName(tableName);
        }

        // Start the threads
        List<Thread> threads = new ArrayList<>();
        for(int i = 0; i < Finals.DB_SHARD_NUM; ++i){
            threads.add(new Thread(workers.get(i)));
            threads.get(i).start();
        }

        // Wait for them all to finish
        for(int i = 0; i < Finals.DB_SHARD_NUM; ++i){
            try{
                threads.get(i).join();
            }
            catch(InterruptedException e){
                TSL.get().err("Facilitator.letWorkersFly InterruptedException");
                e.printStackTrace();
            }
        }
    }

    public void createDBIndex(String columnName, String indexName){
        List<Thread> workers = new ArrayList<>();
        List<Connection> conns = new ArrayList<>();
        String sql = DBCommon.getDBIndexSQLStatement(this.tableName, columnName, indexName);
        if(this.DBAbsolutePaths == null)
            this.DBAbsolutePaths = getDBAbsolutePaths();
        for(String db : this.DBAbsolutePaths)
            conns.add(DBCommon.connect(db));

        for(int i = 0; i < conns.size(); ++i){
            workers.add(new Thread(new IndexWorker(conns.get(i), sql)));
            workers.get(i).start();
        }

        try{
            for(int i = 0; i < workers.size(); ++i)
                workers.get(i).join();
        }
        catch(InterruptedException e){
            e.printStackTrace();
        }

        for(Connection conn : conns)
            DBCommon.disconnect(conn);
    }
}

File: src/main/kotlin/edu/gbcg/dbInteraction/dbcreator/Facilitator.kt
/*
 * Copyright (c) 2018 Sean Grimes. All Rights Reserved.
 * License: MIT
 */

package edu.gbcg.dbInteraction.dbcreator

import edu.gbcg.configs.Finals
import edu.gbcg.dbInteraction.DBCommon
import edu.gbcg.dbInteraction.DBWorker
import edu.gbcg.utils.FileUtils
import edu.gbcg.utils.TSL
import java.sql.Connection
import java.text.NumberFormat

@Suppress("LeakingThis", "ConvertSecondaryConstructorToPrimary")
abstract class Facilitator {
    protected var dbAbsolutePaths_: List<String>        // Path to the DBs once they exist
    protected val dbDirectoryPaths_: List<String>       // Path to the directory / directories that hold the DB shards
    protected val columnNames_: List<String>            // Names of the columns in the DB
    protected val dataTypes_: List<String>              // Type of data stored in the DB columns
    protected val dbPaths_: List<String>                // Paths to the DBs when they don't yet exist
    protected var dataAbsolutePaths_: List<String>      // Paths to the data files
    protected val tableName_: String                    // The name of the table in the DB
    protected val logger_: TSL = TSL.get()              // Instance of the logger
    protected val numberFormat_: NumberFormat           // Format number output for easier viewing
    protected val dataNamesOfInterest_: List<String>    // Name of csv column of json key we need

    constructor(){
        this.dbAbsolutePaths_               = getDBAbsolutePaths()
        this.dbDirectoryPaths_              = getDBDirectoryPaths()
        this.columnNames_                   = getColumnNames()
        this.dataTypes_                     = getDataTypes()
        this.dbPaths_                       = buildDBPaths()
        this.dataAbsolutePaths_             = getDataFileAbsolutePaths()
        this.tableName_                     = getTableName()
        this.numberFormat_                  = NumberFormat.getInstance()
        this.numberFormat_.isGroupingUsed   = true
        this.dataNamesOfInterest_           = getDataKeysOfInterest()
    }

    protected abstract fun buildDBPaths(): List<String>
    protected abstract fun getDataFileAbsolutePaths(): List<String>
    protected abstract fun getDBAbsolutePaths(): List<String>
    protected abstract fun getDBDirectoryPaths(): List<String>
    protected abstract fun getDataTypes(): List<String>
    protected abstract fun getColumnNames(): List<String>
    protected abstract fun getTableName(): String
    protected abstract fun createIndices()
    protected abstract fun dropIndices()
    protected abstract fun getDataAbsolutePathsForNewData(): List<String>
    protected abstract fun getDataKeysOfInterest(): List<String>

    abstract fun pushDataIntoDBs()
    // Creates a list of DataPushers (json, csv, other), passes a list of lines that pusher is supposed to parse
    // and push into the DB, starts each pusher in a thread, lets them work, joins them.
    protected abstract fun letWorkersFly(lines: List<List<String>>)

    // Used by pushDataIntoDBs(), regardless of data type, to check if the function should continue.
    // It will shortcirtuit pushData if the data already exists or START_FRESH is false and ADD_NEW_DATA is false
    // NOTE: The idea really is the stuff going on in this function needs to be enforced across all sub
    // Facilitators so this fucntion moves these important checks into the base Facilitator class
    protected fun shouldFunctionReturnEarly(): Boolean {
        // Early exit if we're not pushing data into the DBs
        if(!Finals.START_FRESH && !Finals.ADD_NEW_DATA) return true
        if(this.dbAbsolutePaths_.isEmpty())
            this.dbAbsolutePaths_ = getDBAbsolutePaths()

        if(this.dataAbsolutePaths_.isEmpty()){
            when{
                Finals.START_FRESH -> this.dataAbsolutePaths_ = getDataFileAbsolutePaths()
                Finals.ADD_NEW_DATA -> this.dataAbsolutePaths_ = getDataAbsolutePathsForNewData()
                else -> logger_.logAndKill("Facilitator.shouldFunctionReturnEarly -- Not START_FRESH or ADD_NEW_DATA")
            }
        }

        // Used for logging when adding new data to the DB shards
        if(Finals.ADD_NEW_DATA)
            this.dataAbsolutePaths_.forEach{logger_.info("Pulling new data from $it")}

        return false
    }

    fun createDBs() {
        if(!Finals.START_FRESH)
            logger_.logAndKill("Called createDBs when Finals.START_FRESH was false. Check your logic.")

        // Check if the DBs exist.
        if(this.dbAbsolutePaths_.isEmpty())
            this.dbAbsolutePaths_ = ArrayList()
        val dbs_exist = this.dbAbsolutePaths_.size == Finals.DB_SHARD_NUM

        // The DBs exist but we want to start fresh, get rid of them
        if(dbs_exist){
            logger_.warn("DB shards exist && starting fresh")
            logger_.info("Dropping table(s) in DB shards")
            val sql = "drop table if exists ${this.tableName_};"
            // For each db in the list, drop the table
            this.dbAbsolutePaths_.forEach{ DBCommon.delete(it, sql) }
        }

        // The DBs don't exist, create the empty sqlite files
        else if(!dbs_exist){
            logger_.info("Creating DB shard paths and initializing DB files")
            // Create the directories that hold the DBs
            this.dbDirectoryPaths_.forEach{ FileUtils.get().checkAndCreateDir(it) }
            // Build the paths to the Dbs so they can be created
            this.dbPaths_.forEach{
                val conn = DBCommon.connect(it)
                DBCommon.disconnect(conn)
            }
            // Now that the DBs exist, populate the absolute paths list
            this.dbAbsolutePaths_ = getDBAbsolutePaths()
        }
        else
            logger_.logAndKill("createDBs; somehow the DBs exist and don't exist. What's that cat joke?")

        // Create the table schema
        val sb = StringBuilder()
        sb.append("create table if not exists ${this.tableName_}(")
        for(i in 0 until this.columnNames_.size){
            sb.append(this.columnNames_[i])
            sb.append(this.dataTypes_[i])
        }
        sb.append(");")
        val sql = sb.toString()

        logger_.info("Writing table to DB shards")
        this.dbAbsolutePaths_.forEach{ DBCommon.insert(it, sql) }
        logger_.info("DBs have been created")
    }

    fun createDBIndex(columnName: String, indexName: String) {
        logger_.info("Creating $indexName on column $columnName")
        val workers = ArrayList<Thread>()
        val conns = ArrayList<Connection>()
        val sql = DBCommon.getDBIndexSQLStatement(this.tableName_, columnName, indexName)
        if(this.dbAbsolutePaths_.isEmpty())
            this.dbAbsolutePaths_ = getDBAbsolutePaths()

        // For each db in dbAbsolutePaths_ create a connection and put it in conns
        this.dbAbsolutePaths_.mapTo(conns) { DBCommon.connect(it) }

        for(i in 0 until conns.size){
            workers.add(Thread(DBWorker(conns[i], sql)))
            workers[i].start()
        }
        try{
            for(i in 0 until workers.size)
                workers[i].join()
        }
        catch(e: InterruptedException){
            logger_.exception(e)
        }

        conns.forEach{ DBCommon.disconnect(it) }
    }

    fun dropDBIndices(indexName: String){
        logger_.info("Dropping $indexName from ${this.tableName_}")
        // Get a connection to each DB shard
        val conns = ArrayList<Connection>()

        // The DB workers, run it in parallel
        val workers = ArrayList<Thread>()

        // Get a connection to each DB and put the connections in conns list
        this.dbAbsolutePaths_.mapTo(conns){DBCommon.connect(it)}

        // Get the SQL statement
        val sql = DBCommon.getDropDBIndexSQLStatement(indexName)

        // Launch the workers
        for(i in 0 until conns.size){
            workers.add(Thread(DBWorker(conns[i], sql)))
            workers[i].start()
        }
        try{
            for(i in 0 until workers.size)
                workers[i].join()
        }
        catch(e: InterruptedException){
            logger_.exception(e)
        }

        // Close all the connections
        conns.forEach{ DBCommon.disconnect(it) }
    }

    // Push new data into the DB shards
    fun pushNewData(){
        // Batch inserting large data on tables with a lot of indices is slow. Drop them and re-create at the end
        dropIndices()

        // Clear the existing list. Note: clear can't be called on a "List" so just replace it with a new one
        this.dataAbsolutePaths_ = ArrayList()

        // Get the path(s) to the new json file(s)
        this.dataAbsolutePaths_ = getDataAbsolutePathsForNewData()

        // Now that the paths have been reset the new data can be pushed in the DB shards
        pushDataIntoDBs()
    }
}
